doc/ni_by_example_4.md

```diff
@@ -1,169 +1,763 @@
-    * The index of the row that the input data that came from
-    * The index of the column that the input data came from
-    * The value of the input stream at the row + column specified by the first two columns.
-* `X` - sparse-to-dense transformation
-  * `X` inverts `Y`; it converts a specifically-formatted 3-column stream into a multiple-column stream.
-  * The specification for what the input matrix must look like is described above in the `Y` operator.
-  
-  
+#`ni` by Example Chapter 4 (alpha release)
+Welcome to Chapter 4. At this point you have enough skills to read the documentation on your own. As a result, this chapter should read a little briefer because it is focused on introducing you to the possibilities of each operator.
+
+Unlike the other chapters thus far, this chapter has no theme; it's a list of useful operations. This chapter covers some of the interplay between `ni` and `bash`, HDFS joins using `nfu`, The `ni` monitor, cell operations, stream splitting, vertical column operations, sparse matrix operations, Ruby and Lisp operators, ni-specific perl operators, and has another Perl chapter.
+
+
+##Calling`ni` from other programming languages
+
+You've already been using `ni` within bash, so you've always been using `ni` from another langauge. In this section, we'll make the connections more explicit, and briefly discuss how to use `ni` from Ruby.
+
+###`ni` and bash
+
+[Spencer](https://github.com/spencertipping) refers to `ni` as Huffman-encoded bash, but we haven't given the treatment of `ni` and bash fully yet. If you're already familiar with bash, this will likely be review.
+
+`e` also can use brackets rather than quotes to execute commands. However, this exposes the code that would have been quoted to bash, which might do something you don't want it to.
+
+When you run a script without quotes, bash will first scan for metacharacters. So the script
+
+```
+$ ni e[seq 10 | grep 1]
+```
+
+is run as 
+
+```
+~/bin/ni "e[seq" "10" | /bin/grep "1]"
+```
+
+because bash will execute the pipe first.
+
+The correct way to execute the script: 
+
+```
+$ni e'seq 10 | grep 1'
+``` 
+
+turns into 
+```
+exec("/bin/sh", "-c", "stuff")
+```
+
+This is a non-obvious feature of the bracketed version of `e`: `e[ word1 word2 ... wordN ]` turns into `exec("word1", "word2", ..., "wordN")`. You'll get shell character expansion with quotes, but not with brackets, the idea being that if you're using brackets, bash has already had a chance to expand the metacharacters like `$foo`.
+
+###`ni` and Ruby
+Aside from bash, Ruby feels like the best scripting language from which to call `ni`. Ruby has an easy and simple syntax for calling shell commands using backticks (i.e. `` `cmd` `` will run using the shell). One thing to be aware of is that Ruby backticks will execute using `/bin/sh` and not `/bin/bash`, so to execute your `ni` spells from Ruby, you will want to create them as strings, and execute them with `bash -c "#{cmd}"`
```

Comment:
That's interesting, because one of the few scripts I've written that use `ni` didn't work in backticks and would with `bash -c #{cmd}`